[16.0] [MIG] eater

victor-champonnois opened the pull request:
internal task : https://gestion.coopiteasy.be/web#id=10884&action=475&active_id=492&model=project.task&view_type=form&menu_id=

I need technical input on two questions  : bca51764f593b3b9acd9bbd03e3b4e63da810fbb and 9902d8890c607fc0219cfe9494528566f8fec883

@polchampion Since the field "customer" has been removed in v13, I removed 
- the restriction on customer in the domain for adding new eaters.
- The invisibility of the field for non-customer contact : so the field Eater/worker is now visible on all contacts

This is not ideal, but I don't see a better solution. We could use the customer_rank field, which starts at zero and increases with the contact's orders. But this would mean that new contacts or contact that never shopped could not been added as Eater.  
Maybe we could make the field invisible for companies ?

victor-champonnois committed:
[REF] split beesdoo_product

into:
- product_main_supplier
- sale_suggested_price
- sale_adapt_price_wizard

[REF] split beesdoo_product -> product_main_supplier

[FIX] sale_suggested_price dep on product_main_supplier

[ADD] module sale_edit_price_wizard

[REF] rename sale_edit_price_wizard ->  sale_adapt_price_wizard

[REF] move rounding_method to sale_suggested_price

[REF] refactor following PR 320

[REF] search product_code -> product_main_seller

[FIX] bug adapt_sales_price

[IMP] use float_compare to compare prices

[REF] move list_price_write_date to sale_adapt_price_wizard

[FIX] beesdoo_product: main_seller_id search

[FIX] run precommit

fix

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

victor-champonnois committed:
[REF] split beesdoo_base : fix dependencies

[REF] put splitted modules as dependencies

[FIX] test-requirements

[FIX] remove beesdoo_base demo data in manifest

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

victor-champonnois committed:
[REF] split beesdoo_base -> eater

[UPD] Update eater.pot

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

victor-champonnois committed:
[REM] deprecated functionalities

[FIX] contributors

[FIX] remove deprecated demo data

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

victor-champonnois committed:
[REF] split beesdoo_base -> eater_member_card

[UPD] Update eater_member_card.pot

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

victor-champonnois committed:
[REF] split beesdoo_base -> member_card

[UPD] Update member_card.pot

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

robinkeunen committed:
[REF] beesdoo_base: remove view renames

We do not need to rename view names when moved from one
module to the other. Odoo cleans up by himself as long
as they are not defined as noupdate.

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

robinkeunen committed:
[REF] eater: rename eater.new.wizard

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

robinkeunen committed:
[REF] eater: dependencies to point_of_sale, purchase

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

robinkeunen committed:
[REF] member_card: rename membercard.new.wizard

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

robinkeunen committed:
[IMP] generate POT and PO files

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

robinkeunen committed:
[IMP] update french translations

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

github-grap-bot committed:
[UPD] README.rst

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

robinkeunen committed:
[IMP] translations

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

huguesdk committed:
[FIX] fix fr translation

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

github-grap-bot committed:
[UPD] README.rst

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[FIX] eater: Add restraints on parent and child eaters

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[FIX] eater: Migration script

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[IMP] eater: Set customer=True for child and parent eaters

This was not correctly restrained previously, so let's fix this.

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

victor-champonnois committed:
[MIG] eater: Migration to 13.0

- remove the customer fields (and domains), because customer is deprecated in Odoo 13.
see here : https://www.odoo.com/forum/help-1/how-to-identify-contact-as-supplier-or-customer-since-v13-185930
We could use a domain on customer_rank and set the rank to 1 for new eaters, but that would
cause issues when the eater is created via standard contact creation.
The customer domain doesn't seem necessary anyway.
- remove old assets

victor-champonnois committed:
[MIG] eater: Migration to 14.0

victor-champonnois committed:
[MIG] eater: Migration to 15.0

victor-champonnois committed:
[MIG] eater: Migration to 16.0

victor-champonnois committed:
[FIX] eater: condition on parent_eater_id

victor-champonnois committed:
[ADD] suggestion to improve performance

victor-champonnois committed:
[ADD] improve comment and propose removal

victor-champonnois committed:
[ADD] eater: tests

polchampion commented:
@victor-champonnois I think it's fine as you did, let's keep it simple. Can you ping me when it's live in test?

cathLemb commented:
"Maybe we could make the field invisible for companies ?"
Not possible as bees for example uses eater field on companies.

carmenbianca commented:
I just realised that #516 is not is not included in this PR, and that the changes to customer will conflict with some changes. I'll fix up #516 and then rebase this work on top of that later down the line.

carmenbianca committed:
[REF] eater: Undo changes to customer

In 13.0, the customer field disappears. To ease the migration, let's
just get rid of the customer requirement already. It didn't add much,
and deleting code is good dopamine.

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca commented on eater/static/font/Roboto-Bold.ttf:
This font (and some other static resources) are deleted in the 13.0 migration commit. Why?

victor-champonnois replied:
I could have removed it in a separate commit. It's just that it's not used and it makes no sense to keep this file here.

carmenbianca commented on eater/views/partner.xml:
```diff
@@ -10,7 +10,7 @@
                 <page
                     string="Eaters"
-                    attrs="{'invisible': ['|', ('customer', '=', False), ('eater', 'not in', ('worker_eater'))]}"
+                    attrs="{'invisible': [('eater', 'not in', ('worker_eater'))]}"
```
This should be in the 13.0 migration commit.

victor-champonnois replied:
done

carmenbianca commented on eater/views/partner.xml:
```diff
@@ -37,7 +37,7 @@
 
             <field name="website" position="after">
-                <field name="eater" attrs="{'invisible': [('customer', '=', False)]}" />
+                <field name="eater" />
```
idem

victor-champonnois replied:
done

carmenbianca commented on eater/models/partner.py:
```diff
@@ -26,7 +26,7 @@ def write(self, values):
                 values.get("parent_eater_id")
                 and partner.parent_eater_id
-                and partner.parent_eater_id.id != values.get("parent_eater_id")
+                and partner.parent_eater_id != values.get("parent_eater_id")
```
This makes no sense to me.

carmenbianca commented on eater/models/partner.py:
```diff
@@ -22,6 +22,8 @@ class Partner(models.Model):
     )
 
     def write(self, values):
+        # fixme: shouldn't the check on "values.get("parent_eater_id")" be before
+        # the for loop to avoid looping through all the contacts at every write?
```
Sure.

victor-champonnois replied:
done

carmenbianca commented on eater/models/partner.py:
```diff
@@ -37,8 +37,10 @@ def write(self, values):
+        # rather than delete
         if "child_eater_ids" in values:
             for command in values["child_eater_ids"]:
                 if command[0] == 2:
```
I guess it depends on why this exists. The apparent goal to me is 'don't accidentally delete the linked child eaters when severing the link'. But why would the '2' command ever be sent when the '3' command is intended?

For context:

- `(2, id, _)` removes the link to and deletes the id related record.
- `(3, id, _)` removes the link to, but does not delete, the id related record.
  This is usually what you will use to delete related records on many-to-many
  fields.

---

In any case, if we keep this line, a test should be added.

victor-champonnois replied:
I would be in favor of removing it then. I don't see why this would be needed for this specific many2many field. Do you agree ?

carmenbianca replied:
It's not a Many2many field. Not sure why the comment calls it a 'many2many command'.

I'm happy to drop this code also. Let's just make a note of it somewhere that the functional tester can be aware of this.

victor-champonnois replied:
done

victor-champonnois replied:
(note in the task)

carmenbianca reviewed (changes requested):
some notes

victor-champonnois committed:
[IMP] perf imp: check for parent eater before loop

victor-champonnois committed:
[REM] no-delete on  child_eater_ids many2many

This check is not necessary. In a many2many widget, removing an item is a "3" (unlink) command. There doesn't seem to be a case where it would be a "2" command.

robinkeunen commented:
The migration commits should be stashed together.

huguesdk commented:
isn’t the module name too generic for such a specific module?

Commit:
[UPD] Update eater.pot

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

github-grap-bot committed:
eater 12.0.2.1.0

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca committed:
[FIX] eater: Improve English, fix check

The check wasn't quite right (it compared the recordset to an id).

The English was also really bad. Patched that up.

Signed-off-by: Carmen Bianca BAKKER <carmen@coopiteasy.be>

carmenbianca commented on eater/models/partner.py:
```diff
@@ -0,0 +1,81 @@
+                )
+
+    def write(self, values):
+        if values.get("parent_eater_id"):
+            for partner in self:
+                if (
+                    partner.parent_eater_id
+                    and partner.parent_eater_id.id != values.get("parent_eater_id")
+                ):
+                    raise ValidationError(
+                        _(
+                            "You are trying to assign %(eater)s as an eater to"
+                            " a new partner, but this eater is already"
+                            " assigned to %(old_partner)s. Please remove the link"
+                            " before creating a new one."
+                        )
+                        % {
+                            "eater": partner.name,
+                            "old_partner": partner.parent_eater_id.name,
+                        }
+                    )
```
What is the use case here?

robinkeunen replied:
You can't be the eater of two workers.

carmenbianca replied:
But this would change the parent eater ("worker"), no? So after the operation, you still only have one parent.

robinkeunen replied:
Indeed, that's weird.

robinkeunen replied:
Apparently, 8 years ago, this field was a many2many => removing the code.

robinkeunen replied:
It dates back from 8 years ago. Let's trust ourselves and state that this is useless.  https://github.com/beescoop/Obeesdoo/commit/3f4d9f69c77c0bd9927e33b34182effb56d46568

carmenbianca reviewed (changes requested):
I rebased this on top of the latest 12.0 to also include #516. I also applied a fix at the end, but I'm not entirely sure why the code that I fixed exists.

I'd like a confirmation that the tagged code is _useful_ before we merge this.

robinkeunen commented on eater/tests/test_eater.py:
```diff
@@ -6,15 +6,15 @@ class TestEater(TransactionCase):
         self.assertTrue(
             self.env.ref("eater.eater2").write(
-                {"parent_eater_id": self.env.ref("eater.eater1")}
+                {"parent_eater_id": self.env.ref("eater.worker1").id}
```
it's better practice to create records in the tests and leave demo data for demo.

robinkeunen commented on eater/demo/eaters.xml:
```diff
@@ -35,4 +35,20 @@
         <field name="is_company" eval="False" />
     </record>
+
+    <record id="worker1" model="res.partner">
+        <field name="name">Gijs Eaterdam</field>
+        <field name="firstname">Gijs</field>
+        <field name="lastname">Eaterdam</field>
+        <field name="eater">worker_eater</field>
+        <field name="is_company" eval="False" />
+    </record>
```
It's better to overload existing partners than creating new ones.

robinkeunen committed:
[IMP] eater: do not display tab if partner not eater/worker

robinkeunen committed:
[REF] eater: do not raise validation error on parent worker update

carmenbianca reviewed (approved):
LGTM @robinkeunen

robinkeunen commented:
/ocabot merge nobump
